### PR TITLE
Improve logging macros

### LINKED ISSUE #52074: The macros related to LOG_IMPL will eval macro parameters (i.e. __VA_ARGS__) twice or more
Hi, I found that the LOG_XXX macros (Common/logger_userful.h) will evaluate `__VA_ARGS__` twice after expanded.

https://github.com/ClickHouse/ClickHouse/blob/9df928eb132dcd99e446f9dc7897ea613f7c64a7/src/Common/logger_useful.h#L44C4-L44C4

e.g. 
``` C++
// the expr `a + b` will be evaluated twice
LOG_DEBUG(log, "echo {}", a + b); 

// key line after expanded : std::string formatted_message = numArgs(a+b) ? fmt::format(a+b) : firstArg(a+b);

```

the reason is that expr `a+b` will be evaluated as the parameter of numArgs, and soon be evaluated as the parameter of fmt::format.

NOTE: `constexpr` on numArgs won't prevent the evaluation of parameters (try `consteval`, you'll find the parameters of numArgs cannot be folded in compile time). And release version won't optimize this yet.

--- 

Also, if  a function with side effects used here, **potential bug** might be introduced, like this
``` C++
// some functions with side effects might be called inside LOG_XXX macros.
int foo() { std::cout << "foo" << std::endl; return 1; }
// text foo will be printed twice
LOG_DEBUG(log, "echo {}", foo());

// variable `sth` will be moved in the first expanded expr and will be empty in the second expanded expr
LOG_DEBUG(log, "echo {}", toString(std::move(sth))); 

// key line after expanded : std::string formatted_message = numArgs(toString(std::move(sth))) ? fmt::format(toString(std::move(sth))) : firstArg(toString(std::move(sth)));
```

I tried a possible solution that works. 
``` C++
template <typename... Ts>
auto getMessage(fmt::format_string<Ts...>&& fs, Ts&&... ts) {
  if constexpr (sizeof...(Ts) == 0) {
    fmt::string_view sv(fs);
    return std::string(sv.data(), sv.size());
  } else {
    return fmt::format(fs, std::forward<Ts>(ts)...); /// requires >= fmt 8.0
  }
}

std::string formatted_message = getMessage(__VA_ARGS__);
```

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fixes https://github.com/ClickHouse/ClickHouse/issues/52074

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 1ceaa0d4ac462cca83bd5558a05129b00a654c7b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52519/1ceaa0d4ac462cca83bd5558a05129b00a654c7b/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by alexey-milovidov — APPROVED** [review]

**Review by nickitat — APPROVED** [review]

File: src/Common/logger_useful.h [hunk 1]
```diff
@@ -28,33 +28,86 @@ namespace
...
+            static_assert(_nargs == 1 || !is_preformatted_message);                                                 \
+            ConstexprIfsAreNotIfdefs<is_preformatted_message>::getPreformatted(LOG_IMPL_FIRST_ARG(__VA_ARGS__)).apply(_formatted_message, _format_string);  \
+        }                                                                                                           \
+        else                                                                                                        \
+        {                                                                                                           \
+             _formatted_message = _nargs == 1 ? firstArg(__VA_ARGS__) : fmt::format(__VA_ARGS__);                   \
+        }                                                                                                           \
+                                                                                                                    \
+        std::string _file_function = __FILE__ "; ";                                                                 \
```
Comment (nickitat) [inline src/Common/logger_useful.h:101]: we prob could replace it with `source_location`, but I don't know for what

Comment (tavplubix) [reply src/Common/logger_useful.h:101]: Last time I checked it didn't work correctly in some cases

**Comment by tavplubix** [comment]
AST fuzzer (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/52632
Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/41850

**Pull request merged**